src/App.tsx

```diff
@@ -26,70 +27,135 @@ type ReorderOptions = {
-          type="button"
-          className="button is-warning is-light"
-        >
-          Reverse
-        </button>
-
-        <button
-          type="button"
-          className="button is-danger is-light"
-        >
-          Reset
-        </button>
-      </div>
+type State = {
+  isReversed: boolean,
+  sortType: SortType,
+};
+
+export class App extends React.Component<{}, State> {
+  state = {
+    isReversed: false,
+    sortType: SortType.NONE,
+  };
+
+  reverse = () => {
+    this.setState(state => ({
+      isReversed: !state.isReversed,
+    }));
+  };
+
+  sortByType = (type: SortType) => {
+    this.setState({ sortType: type });
+  };
+
+  reset = () => {
+    this.sortByType(SortType.NONE);
+    this.setState({ isReversed: false });
+  };
+
+  render() {
+    const { isReversed, sortType } = this.state;
+    const showResetButton = sortType !== SortType.NONE || isReversed;
+    const reorderedGoods = getReorderedGoods(
+      goodsFromServer, this.state,
+    );
+
+    return (
+      <div className="section content">
+        <div className="buttons">
+          <button
+            type="button"
+            className={classNames(
+              'button is-info', {
+                'is-light': sortType !== SortType.ALPHABET,
+              },
+            )}
+            onClick={() => {
+              this.sortByType(SortType.ALPHABET);
+            }}
+          >
+            Sort alphabetically
+          </button>
+
+          <button
+            type="button"
+            className={classNames(
+              'button is-success', {
+                'is-light': sortType !== SortType.LENGTH,
+              },
+            )}
+            onClick={() => {
+              this.sortByType(SortType.LENGTH);
+            }}
+          >
+            Sort by length
+          </button>
+
+          <button
+            type="button"
+            className={classNames(
+              'button is-warning', {
+                'is-light': !isReversed,
+              },
+            )}
+            onClick={() => {
+              this.reverse();
+            }}
+          >
+            Reverse
+          </button>
+
+          {showResetButton && (
+            <button
+              type="button"
+              className="button is-danger is-light"
+              onClick={() => {
+                this.reset();
+              }}
+            >
+              Reset
+            </button>
+          )}
+        </div>
 
-      <ul>
         <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
+          <ul>
+            {
+              reorderedGoods.map(good => (
+                <li data-cy="Good" key={good}>
+                  {good}
+                </li>
+              ))
+            }
+          </ul>
         </ul>
```

Comment:
Do you really need two `ul` elements here?

```suggestion
          {
            reorderedGoods.map(good => (
              <li data-cy="Good" key={good}>
                {good}
              </li>
            ))
          }
        </ul>
```